cmd/control_loop/controller/controller.go

```diff
@@ -118,38 +122,39 @@ func updateResources(
 
 	// TODO: Add more error handling in the two loops
 	for _, resource := range resourceList {
-		r_ts := resource.GetUpdateTimestamp()
+		resourceTime := resource.GetUpdateTimestamp()
+
 		takeAction := false
-		var args []string
+		var args []resources.Resource
 
 		// Evaluate this resource against each dependency source pattern
 		for i, dependency := range dependencies {
 			dMap := dependencyMaps[i]
 			// Get the group to look for in dependencyMap
-			groupFuncName := ParseGroupFunc(dependency.Source)
-			groupBy := ""
-
-			if len(groupFuncName) > 0 {
-				groupFunc := reflect.ValueOf(resource).MethodByName(groupFuncName)
-				groupBy = groupFunc.Call([]reflect.Value{})[0].String()
+			group, err := ExtractGroup(dependency.Source, resource)
+			if err != nil {
+				return nil, errors.New(fmt.Sprintf("Cannot match resource with source. Error: %s", err.Error()))
 			}
 
-			if collection, ok := dMap[groupBy]; ok {
+			if collection, ok := dMap[group]; ok {
 			// Take action if dependency timestamp is later than resource timestamp
-				if (*collection).maxUpdateTime.After(r_ts) {
+				if (*collection).maxUpdateTime.After(resourceTime) {
 					takeAction = true
 				}
-				visited[groupBy] = true
-				argVal := DeriveArgsFromResources(i, (*(*collection).resourceList)[0], action)
-				args = append(args, argVal)
+				visited[group] = true
+				args = append(args, (*(*collection).resourceList)[0])
 			} else {
 				takeAction = false
 				break
```

Comment:
`takeAction=false` exists here to avoid miscalculating an action. 

I can explain this scenario with an example:
Let's say our resource is `apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/score`
We have two dependencies:
1. `apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/lint-gnostic`
2. `apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/complexity`

What this loop is essentially doing is, 1. comparing timestamps and 2. checking if all of the dependency groups are present. To correctly update the `score` artifact in the above example, both of it's dependencies `lint-gnostic` and `complexity` should be present. If either of those is not present, we totally avoid generating a command for it and just break out of the loop.

If you look at `GenerateCommand()` it expects `args` to be exactly of same length as the dependencies. Hence, when at least one of the dependencies is not present, we do `takeAction=false` and avoid the case of `GenerateCommand` calculating a wrong action.

Makes sense?